lib/manager/components/deployment/CustomFileEditor.js

```diff
@@ -0,0 +1,256 @@
+      model: {
+        ...this.state.model,
+        filename: evt.target.value
+      }
+    })
+  }
+
+  _onChangeServeUse = () => {
+    const { model } = this.state
+    this.setState({
+      model: {
+        ...model,
+        useDuringServe: !model.useDuringServe
+      }
+    })
+  }
+
+  _onChangeSource = (evt: SyntheticInputEvent<HTMLSelectElement>) => {
+    const model = {...this.state.model}
+    // set variable to make flow happy
+    let newSource
+    if (evt.target.value === 'raw') {
+      model.uri = null
+      newSource = 'raw'
+    } else {
+      model.contents = null
+      newSource = 'uri'
+    }
+    this.setState({
+      fileSource: newSource,
+      model
+    })
+  }
+
+  _onChangeUri = (evt: SyntheticInputEvent<HTMLInputElement>) => {
+    this.setState({
+      model: {
+        ...this.state.model,
+        uri: evt.target.value
+      }
+    })
+  }
+
+  _onCancelEditing = () => {
+    const { customFile, onCancelEditing } = this.props
+    this.setState({
+      fileSource: customFile.contents ? 'raw' : 'uri',
+      model: customFile
+    })
+    onCancelEditing()
+  }
+
+  _onDelete = () => {
+    const { idx, onDelete } = this.props
+    onDelete(idx)
+  }
+
+  _onEdit = () => {
+    const { idx, onEdit } = this.props
+    onEdit(idx)
+  }
+
+  _onSave = () => {
+    const { idx, onSave } = this.props
+    onSave(idx, this.state.model)
+  }
+
+  _renderToolbar = () => {
+    const isEditing = this._isEditing()
+    const canEdit = this._canEdit()
+    const either = !isEditing && !canEdit
+    return (
+      <ButtonToolbar>
+        <Button
+          bsSize='xsmall'
+          disabled={either}
+          onClick={canEdit ? this._onEdit : this._onCancelEditing}
+        >
+          {isEditing ? 'Cancel' : 'Edit'}
+        </Button>
+        {isEditing &&
+          <Button
+            bsSize='xsmall'
+            disabled={!this._isValid()}
+            onClick={this._onSave}
+          >
+            Save
+          </Button>
+        }
+        <Button
+          bsSize='xsmall'
+          bsStyle='danger'
+          className='pull-right'
+          disabled={either}
+          onClick={this._onDelete}
+        >
+          Delete
+        </Button>
+      </ButtonToolbar>
+    )
+  }
+
+  render () {
+    const {
+      fileSource,
+      model: customFile
+    } = this.state
+    const isValid = this._isValid()
+    const isEditing = this._isEditing()
+    return (
+      <div className='custom-file'>
+        {this._renderToolbar()}
+        <div className='margin-top-15'>
+          {isEditing
+            ? <FormControl
+              onChange={this._onChangeFilename}
+              placeholder='Enter filename'
+              type='text'
+              value={customFile.filename}
+            />
+            : <span>
+              Filename:{' '}
+              <span style={{fontFamily: 'monospace'}}>
+                {/* FIXME: what should the behavior be here if missing */}
```

Comment:
There should be a validation error if the filename is missing when raw contents are provided. I will fix this in my PR.